python/lsst/pipe/base/_check_qg_outputs.py

```diff
@@ -0,0 +1,216 @@
+            if output_ref.id not in refs[output_ref.datasetType.name]:
+                dataset_type_report.handle_missing_dataset(output_ref, failed, status_graph)
+            else:
+                dataset_type_report.handle_produced_dataset(output_ref, status_graph)
+
+    def to_summary_dict(self, butler: Butler, logs: bool = True) -> dict[str, Any]:
+        failed_quanta = {}
+        for node_id, log_ref in self.failed.items():
+            quantum_info = {"data_id": log_ref.dataId.byName()}
+            if logs:
+                try:
+                    log = butler.get(log_ref)
+                except LookupError:
+                    quantum_info["error"] = []
+                else:
+                    quantum_info["error"] = [
+                        record.message for record in log if record.levelno >= logging.ERROR
+                    ]
+            failed_quanta[str(node_id)] = quantum_info
+        return {
+            "outputs": {name: r.to_summary_dict() for name, r in self.output_datasets.items()},
+            "failed_quanta": failed_quanta,
+            "n_quanta_blocked": len(self.failed_upstream),
+        }
+
+    def __str__(self) -> str:
+        return f"failed: {len(self.failed)}\nfailed upstream: {len(self.failed_upstream)}\n"
+
+
+@dataclasses.dataclass
+class QuantumGraphExecutionReport:
+    tasks: dict[str, TaskExecutionReport] = dataclasses.field(default_factory=dict)
+
+    def to_summary_dict(self, butler: Butler, logs: bool = True) -> dict[str, Any]:
+        return {task: report.to_summary_dict(butler, logs=logs) for task, report in self.tasks.items()}
+
+    def write_summary_yaml(self, butler: Butler, filename: str, logs: bool = True) -> None:
+        with open(filename, "w") as stream:
+            yaml.safe_dump(self.to_summary_dict(butler, logs=logs), stream)
+
+    @classmethod
+    def make_reports(
+        cls,
+        butler: Butler,
+        graph: QuantumGraph | ResourcePathExpression,
+    ) -> QuantumGraphExecutionReport:
+        refs = {}
+        status_graph = networkx.DiGraph()
+        if not isinstance(graph, QuantumGraph):
+            qg = QuantumGraph.loadUri(graph)
+        else:
+            qg = graph
+        collection = qg.metadata["output_run"]
+        report = cls()
+        task_defs = list(qg.iterTaskGraph())
+        pipeline_dataset_types = PipelineDatasetTypes.fromPipeline(task_defs, registry=butler.registry)
+        for dataset_type in itertools.chain(
+            pipeline_dataset_types.initIntermediates,
+            pipeline_dataset_types.initOutputs,
+            pipeline_dataset_types.intermediates,
+            pipeline_dataset_types.outputs,
+        ):
+            refs[dataset_type.name] = {
+                ref.id: ref
+                for ref in butler.registry.queryDatasets(
+                    dataset_type.name, collections=collection, findFirst=False
+                )
+            }
+        for taskDef in qg.iterTaskGraph():
```

Comment:
Since pretty much everything in this file is snake_case, `taskDef` should be, too (also below).